src/js/components/returns/inbound/AddItemsPage.jsx

```diff
@@ -314,61 +343,82 @@ class AddItemsPage extends Component {
       this.props.showSpinner();
       return apiClient.post(updateItemsUrl, flattenRequest(payload))
         .then(() => this.fetchInboundReturn())
-        .catch(() => {
-          this.props.hideSpinner();
-          return Promise.reject(new Error(this.props.translate('react.inboundReturns.error.saveOrderItems.label', 'Could not save order items')));
-        });
+        .catch(() => Promise.reject(new Error(this.props.translate('react.inboundReturns.error.saveOrderItems.label', 'Could not save order items'))))
+        .finally(() => this.props.hideSpinner());
     }
 
     return Promise.reject();
   }
 
-  save(formValues) {
-    const returnItems = _.filter(formValues.returnItems, item => !_.isEmpty(item));
+  async saveStockTransferInCurrentStep(formValues, status = null) {
+    const returnItems = _.filter(formValues.returnItems, (item) => !_.isEmpty(item) && item.product);
 
-    if (_.some(returnItems, item => !item.quantity || item.quantity === '0')) {
-      this.confirmSave(() => this.saveItems(returnItems));
-    } else {
-      this.saveItems(returnItems);
+    const hasEmptyOrZeroValues = _.some(returnItems, (item) => !item.quantity || item.quantity === '0');
+
+    if (hasEmptyOrZeroValues) {
+      const isConfirmed = await this.confirmEmptyQuantitySave();
+      if (!isConfirmed) {
+        return Promise.reject();
+      }
     }
+
+    let isSomeItemsDontMatchExpirationDate = false;
+    const itemsWithLotAndExpirationDate = returnItems.filter(it => it.expirationDate && it.lotNumber);
+
+    // Trying to find at least one instance where the data that we are trying to save
+    // does not match the expiration date of the existing inventoryItem in the system
+    for (const it of itemsWithLotAndExpirationDate) {
+      const { data } = await ProductApi.getInventoryItem(it.product?.id, it.lotNumber);
+      if (data.inventoryItem && data.inventoryItem.expirationDate !== it.expirationDate) {
+        isSomeItemsDontMatchExpirationDate = true;
+        break;
+      }
+    }
+
+    // After find at least a single instance where expiration date we are trying to save
+    // does not match the existing inventoryItem expiration date, we want to inform the user
+    // that certain updates to th expiration date in the system will be performed
+    if (isSomeItemsDontMatchExpirationDate) {
+      const isConfirmed = await this.confirmExpirationDateSave();
+      if (!isConfirmed) {
+        return Promise.reject();
+      }
+    }
```

Comment:
Wouldn't it be better to place this code from `if (isSomeItemsDontMatchExpirationDate) { ... }` directly in `if (data.inventoryItem && data.inventoryItem.expirationDate !== it.expirationDate)`?
Basically, you are trying to find at least one item that doesn't match the expiration date, so I think there is no need to use an additional let declaration, and then check if we change the value of this particular variable because within this loop we already know if we found something that doesn't match expiration date.
So I think it should be something similar to (not tested):
```
  for (const it of itemsWithLotAndExpirationDate) {
      const { data } = await ProductApi.getInventoryItem(it.product?.id, it.lotNumber);
      if (data.inventoryItem && data.inventoryItem.expirationDate !== it.expirationDate) {
       const isConfirmed = await this.confirmExpirationDateSave();
        if (!isConfirmed) {
          return Promise.reject();
        }
      }
  }
```

Reply:
I agree with Alan's comment here